### ISSUE TITLE
[MU4 Issue]  Back-up  never happens

### ISSUE THREAD

**Issue by wizofaus** [musescore/MuseScore#9194]
**Describe the bug**
If MuseScore crashes at any time while trying to save a file, even during AutoSave, the mscz is left unusable, and no recovery appears possible (no backup file created).

**To Reproduce**
1. No single good way to reproduce except using the debugger and setting breakpoints/terminating process part way through saving.  Though it might be possible by using "remove empty trailing measures" before saving.

**Expected behaviour**
Original file should remain untouched until saving has finished, then updated via a rename or copy. Or at least, a backup should be created before saving, so that any unexpected application abort during save doesn't mean losing already saved work.
Though obviously in an ideal world the application should never abort unexpectedly while saving!

**Screenshots**

**Desktop (please complete the following information):**
Windows 10

**Additional context**
BTW while it appears to be coded to create a back-up every time it saves, in fact it never happens as `NotationProject::makeCurrentFileAsBackup()` exits early after determining that there's no valid "created()" property. But even after addressing this, the test `if (fileSystem()->exists(filePath))` after this causes another early exit - I'm pretty sure there's supposed to be a ! in there (i.e. if the file *doesn't* exist!).

**Comment by wizofaus** [comment]
If it didn't crash then you can't have verified that a backup got created (or the file didn't get corrupted).
Unfortunately like I said it's hard to provide a reliable easy of having it crash while saving unless you use a debugger to stop/kill the process.

**Comment by Tantacrul** [comment]
OK, so I've tested something related to this (I don't have the ability to force a save operation to crash)

1. Created a new score. Did not save it. Wrote a few notes. Used the 'Crash' button in Dev Tools to intentionally crash the app.
2. When reopening, I got this message:
<img width="349" alt="image" src="https://user-images.githubusercontent.com/21022311/166326356-6e00885d-1ff4-4cba-9036-405ebb522c1f.png">
3. I then reopened the file and continued writing. Crashed again. Reopened again
4. Since autosave only happens every few mins (I think), I often lost a bit of recently added information, which is fine / to be expected

@wizofaus - can you advise whether you think this should be reopened or whether there is another issue to be raised about this?

**Comment by wizofaus** [comment]
The biggest problem is fixed - it now saves to a temporary file and only replaces the original file at the end, so even a crash during save won't corrupt the mscz file.

But the logic in  NotationProject::makeCurrentFileAsBackup() still looks wrong to me

    Ret ret = fileSystem()->exists(filePath);
    if (ret) {
        LOGE() << "project file does not exist";
        return ret;
    }

Basically that's saying "If the file exists, log that it *doesn't*, then return and don't actually back-up". The logic to *do* the backup only occurs if the file *doesn't* exist, which will necessarily fail.
But worse than that, from the testing I did, it never even reaches *that* point - the check just before that `if (isNewlyCreated())` always seems to return true and causes the function to never do anything.
The upshot of which means that as far as I can tell, it will never create a backup of the "current state" of the file before writing a new one.  The intention seems to be that if I save once, I get "MyScore.mscz", then make some modifications, it will first move "MyScore.mscz" to "MyScore.mscz~" then write a new MyScore.mscz.  But that move step is never happening, so MyScore.mscz~ is never created.

So yes, I'd say that's still an issue.

**Comment by cbjeukendrup** [comment]
It seems like we're talking about two distinct things here: autosave vs backup.

Autosave is that every few minutes a copy of the file is saved to a file like `filename.mscz.autosave`; when reopening MU after a crash, it will detect the autosave file and prompt to restore the session. That works well by now.

But backup is a much less well-known feature, that copies the file to a secret location just before you save it, so that you still have the old version somewhere in case you decide you want to go back (i.e. discard your changes after you have already saved them). This feature is indeed not quite working yet, as @wizofaus is describing.

**Comment by Tantacrul** [comment]
In that case, I'm attaching a P1 and targeting it at Beta. Please (anyone) let me know if they think the priorities should be changed.

**Comment by wizofaus** [comment]
I'd even say P2, as Casper noted, it's not that well known a feature, more or less a poor-man's automatic version control. But your call!

**Comment by Tantacrul** [comment]
(worth mentioning a p1 is lower than a p0)

P0 was Vasily's idea. I'm used to p1's being the highest :)

Anyway, p2. Why not?

**Comment by DmitryArefiev** [comment]
Should work now. And after #13385 it should work like in MuseScore3

Closing 

@wizofaus Please reopen if it still happens
